File: apps/api/src/services/vector.service.ts
import { config } from '../config/env.js';
import { logger } from '../config/logger.js';

export interface VectorPoint {
  id: string;
  vector: number[];
  payload: Record<string, unknown>;
}

export interface SearchResult {
  id: string;
  score: number;
  payload: Record<string, unknown>;
}

export interface VectorCollection {
  name: string;
  vectors: number;
  points: number;
}

class VectorService {
  private baseUrl: string;
  private grpcPort: number;
  private apiKey: string;

  constructor() {
    const host = config.qdrant.host.startsWith('http') 
      ? config.qdrant.host 
      : `https://${config.qdrant.host}`;
    this.baseUrl = `${host}:${config.qdrant.port}`;
    this.grpcPort = config.qdrant.grpcPort;
    this.apiKey = config.qdrant.apiKey;
  }

  private getHeaders() {
    return {
      'Content-Type': 'application/json',
      ...(this.apiKey ? { 'api-key': this.apiKey } : {}),
    };
  }

  private getCollectionName(userId: string): string {
    return `user_${userId}`.replace(/-/g, '_').replace(/:/g, '_');
  }

  async createCollection(name: string, vectorSize: number): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/collections`, {
        method: 'POST',
        headers: this.getHeaders(),
        body: JSON.stringify({
          name,
          vectors: {
            size: vectorSize,
            distance: 'Cosine',
          },
        }),
      });

      return response.ok;
    } catch (error) {
      logger.error({ error, name, vectorSize }, 'Failed to create collection');
      throw error;
    }
  }

  async deleteCollection(name: string): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/collections/${name}`, {
        method: 'DELETE',
        headers: this.getHeaders(),
      });

      return response.ok;
    } catch (error) {
      logger.error({ error, name }, 'Failed to delete collection');
      throw error;
    }
  }

  async upsertPoints(collection: string, points: VectorPoint[]): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/collections/${collection}/points`, {
        method: 'PUT',
        headers: this.getHeaders(),
        body: JSON.stringify({
          points: points.map((p) => ({
            id: p.id,
            vector: p.vector,
            payload: p.payload,
          })),
        }),
      });

      return response.ok;
    } catch (error) {
      logger.error({ error, collection, pointsCount: points.length }, 'Failed to upsert points');
      throw error;
    }
  }

  async searchPoints(
    collection: string,
    vector: number[],
    limit = 10,
    scoreThreshold = 0.0,
    filter?: Record<string, unknown>
  ): Promise<SearchResult[]> {
    try {
      const filterBody = filter
        ? {
            filter: {
              must: Object.entries(filter).map(([key, value]) => ({
                key,
                match: { value },
              })),
            },
          }
        : {};

      const response = await fetch(`${this.baseUrl}/collections/${collection}/points/search`, {
        method: 'POST',
        headers: this.getHeaders(),
        body: JSON.stringify({
          vector,
          limit,
          score_threshold: scoreThreshold,
          ...filterBody,
          with_payload: true,
        }),
      });

      if (!response.ok) {
        throw new Error(`Search failed: ${response.statusText}`);
      }

      const data = await response.json() as any;
      
      return data.result.map((p: { id: string; score: number; payload: Record<string, unknown> }) => ({
        id: p.id,
        score: p.score,
        payload: p.payload,
      }));
    } catch (error) {
      logger.error({ error, collection, vectorSize: vector.length }, 'Failed to search points');
      throw error;
    }
  }

  async deletePoints(collection: string, ids: string[]): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/collections/${collection}/points/delete`, {
        method: 'POST',
        headers: this.getHeaders(),
        body: JSON.stringify({
          points: ids,
        }),
      });

      return response.ok;
    } catch (error) {
      logger.error({ error, collection, ids }, 'Failed to delete points');
      throw error;
    }
  }

  async getCollections(): Promise<VectorCollection[]> {
    try {
      const response = await fetch(`${this.baseUrl}/collections`);
      
      if (!response.ok) {
        throw new Error(`Failed to get collections: ${response.statusText}`);
      }

      const data = await response.json() as any;
      
      return data.result.map((c: { name: string; vectors: number; points: number }) => ({
        name: c.name,
        vectors: c.vectors,
        points: c.points,
      }));
    } catch (error) {
      logger.error({ error }, 'Failed to get collections');
      throw error;
    }
  }

  async collectionExists(name: string): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/collections/${name}`);
      return response.ok;
    } catch {
      return false;
    }
  }

  async checkHealth(): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/readyz`);
      return response.ok;
    } catch {
      return false;
    }
  }

  async updateCollectionMetadata(name: string, metadata: Record<string, unknown>): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/collections/${name}`, {
        method: 'PATCH',
        headers: this.getHeaders(),
        body: JSON.stringify({
          metadata,
        }),
      });

      return response.ok;
    } catch (error) {
      logger.error({ error, name, metadata }, 'Failed to update collection metadata');
      throw error;
    }
  }

  async getCollectionInfo(name: string): Promise<any> {
    try {
      const response = await fetch(`${this.baseUrl}/collections/${name}`);
      if (!response.ok) {
        throw new Error(`Failed to get collection: ${response.statusText}`);
      }
      return response.json();
    } catch (error) {
      logger.error({ error, name }, 'Failed to get collection info');
      throw error;
    }
  }
}

export const vectorService = new VectorService();

File: apps/ai-service/app/services/vector_service.py
from typing import List, Optional, Dict, Any
import uuid
from qdrant_client import QdrantClient
from qdrant_client.models import Distance, VectorParams, Filter, FieldCondition, MatchValue
from app.core.config import get_settings
from app.core.logger import app_logger

settings = get_settings()


class VectorService:
    def __init__(self):
        host = settings.QDRANT_HOST
        if not host.startswith("http"):
            host = f"https://{host}"
        self.client = QdrantClient(
            url=host,
            api_key=settings.QDRANT_API_KEY,
        )
        app_logger.info(f"Vector service initialized: {host}")

    async def create_collection(
        self,
        collection_name: str,
        vector_size: int = 1536,
        distance: Distance = Distance.COSINE,
    ) -> bool:
        try:
            self.client.recreate_collection(
                collection_name=collection_name,
                vectors_config=VectorParams(
                    size=vector_size,
                    distance=distance,
                ),
            )
            app_logger.info(f"Collection created: {collection_name}")
            return True
        except Exception as e:
            app_logger.error(f"Error creating collection {collection_name}: {e}")
            return False

    async def delete_collection(self, collection_name: str) -> bool:
        try:
            self.client.delete_collection(collection_name=collection_name)
            app_logger.info(f"Collection deleted: {collection_name}")
            return True
        except Exception as e:
            app_logger.error(f"Error deleting collection {collection_name}: {e}")
            return False

    async def insert(
        self,
        collection_name: str,
        vectors: List[List[float]],
        payloads: List[Dict[str, Any]],
        ids: Optional[List[str]] = None,
    ) -> bool:
        try:
            points = []
            for i in range(len(vectors)):
                point_id = ids[i] if ids else str(uuid.uuid4())
                points.append({
                    "id": point_id,
                    "vector": vectors[i],
                    "payload": payloads[i],
                })
            self.client.upsert(
                collection_name=collection_name,
                points=points,
            )
            app_logger.info(f"Inserted {len(vectors)} vectors into {collection_name}")
            return True
        except Exception as e:
            app_logger.error(f"Error inserting vectors: {e}")
            return False

    async def search(
        self,
        collection_name: str,
        query_embedding: List[float],
        limit: int = 10,
        score_threshold: float = 0.0,
        filter_conditions: Optional[Dict[str, Any]] = None,
    ) -> List[Dict[str, Any]]:
        try:
            search_filter = None
            if filter_conditions:
                search_filter = Filter(
                    must=[
                        FieldCondition(
                            key=key,
                            match=MatchValue(value=value),
                        )
                        for key, value in filter_conditions.items()
                    ]
                )

            results = self.client.query_points(
                collection_name=collection_name,
                query=query_embedding,
                limit=limit,
                score_threshold=score_threshold if score_threshold > 0 else None,
                query_filter=search_filter,
                with_payload=True,
            )

            result_list = list(results)
            points = []
            if result_list and len(result_list[0]) >= 2:
                points = result_list[0][1]
            
            return [
                {
                    "id": r.id,
                    "score": r.score,
                    "payload": r.payload,
                }
                for r in points
            ]
        except Exception as e:
            app_logger.error(f"Error searching: {e}")
            return []

    async def delete_points(self, collection_name: str, ids: List[str]) -> bool:
        try:
            self.client.delete(
                collection_name=collection_name,
                points_selector=ids,
            )
            return True
        except Exception as e:
            app_logger.error(f"Error deleting points: {e}")
            return False

    async def collection_exists(self, collection_name: str) -> bool:
        try:
            collections = self.client.get_collections().collections
            return any(c.name == collection_name for c in collections)
        except Exception as e:
            app_logger.error(f"Error checking collection: {e}")
            return False


vector_service = VectorService()